File: internal/model/slots.go
package model

import (
	"fmt"
	"regexp"
	"slices"
	"strconv"
	"strings"
	"time"
)

// DeploymentSlot is a second copy of a node or worker site that runs its
// own release on its own instances, like an Azure App Service deployment
// slot: deploy to it, test it on its own bindings (staging.example.com),
// then swap it into production. Production is the site itself and not a
// slot of this list.
//
// A slot runs with production's configuration except for its slot
// settings: its own variables (Env, which override production's and are
// never swapped), production's variables marked SlotSetting (which stay in
// production), its instance count and its bindings (Binding.Slot). Before
// a swap the slot's instances are restarted with production's settings
// and warmed up, so that production traffic moves onto processes that are
// already running with the configuration they will keep.
type DeploymentSlot struct {
	Name string `json:"name"` // "staging": lower-case letters, digits and '-'
	// Env are the slot's own variables: added to production's variables
	// that are not slot settings, replacing any of the same name.
	Env []EnvVar `json:"env,omitempty"`
	// Instances run in this slot; 0 = as many as production. While a
	// swap warms the slot up it runs production's count.
	Instances int          `json:"instances,omitempty"`
	AutoSwap  bool         `json:"autoSwap"` // swap into production after a successful deployment to this slot
	Warmup    WarmupConfig `json:"warmup"`
	// ActiveRelease is the deployment this slot runs, managed by
	// NodeHoster like Site.ActiveRelease. Empty = nothing deployed yet: the
	// slot is not started.
	ActiveRelease string `json:"activeRelease,omitempty"`
}

// WarmupConfig is how a slot's instances are warmed up before a swap,
// like Azure's applicationInitialization: every path is requested on every
// instance until it answers an accepted status or the timeout expires.
type WarmupConfig struct {
	Paths      []string `json:"paths"`      // default "/"
	Statuses   string   `json:"statuses"`   // accepted statuses, e.g. "200-399" (the default) or "200-299,401"
	TimeoutSec int      `json:"timeoutSec"` // for all instances and paths together; default 120
}

const (
	// ProductionSlot names the site itself where a slot name is expected.
	ProductionSlot = "production"
	// MaxSlots is how many slots a site may have besides production.
	MaxSlots = 4

	DefaultWarmupTimeoutSec = 120
	MaxWarmupTimeoutSec     = 1800
	DefaultWarmupStatuses   = "200-399"
)

var slotNameRe = regexp.MustCompile(`^[a-z0-9]([a-z0-9-]{0,30}[a-z0-9])?$`)

// slotKeySep joins a site ID and a slot name. Site IDs are UUIDs and slot
// names cannot contain it.
const slotKeySep = "@"

// SlotKey identifies a slot's processes and traffic: the site ID for
// production, "<site ID>@<slot>" for another slot.
func SlotKey(siteID, slot string) string {
	if slot == "" || slot == ProductionSlot {
		return siteID
	}
	return siteID + slotKeySep + slot
}

// SplitSlotKey is the inverse of SlotKey; slot is "" for production.
func SplitSlotKey(key string) (siteID, slot string) {
	siteID, slot, _ = strings.Cut(key, slotKeySep)
	return siteID, slot
}

// NormalizeSlot maps "production" to "" (the site itself).
func NormalizeSlot(name string) string {
	name = strings.ToLower(strings.TrimSpace(name))
	if name == ProductionSlot {
		return ""
	}
	return name
}

// FindSlot returns the named slot, nil if the site has none by that name.
func (s *Site) FindSlot(name string) *DeploymentSlot {
	for i := range s.Slots {
		if s.Slots[i].Name == name {
			return &s.Slots[i]
		}
	}
	return nil
}

// ReleaseIn is the release a slot runs ("" = production's).
func (s *Site) ReleaseIn(slot string) string {
	if slot == "" {
		return s.ActiveRelease
	}
	if sl := s.FindSlot(slot); sl != nil {
		return sl.ActiveRelease
	}
	return ""
}

// SlotReleases are the releases the site's slots run, which deployments
// must not prune.
func (s *Site) SlotReleases() []string {
	var out []string
	for _, sl := range s.Slots {
		if sl.ActiveRelease != "" {
			out = append(out, sl.ActiveRelease)
		}
	}
	return out
}

// SlotBindings are the bindings that route to a slot ("" = production).
func (s *Site) SlotBindings(slot string) []Binding {
	out := []Binding{}
	for _, b := range s.Bindings {
		if b.Slot == slot {
			out = append(out, b)
		}
	}
	return out
}

// HasSlotBindings reports whether any binding routes to a slot rather than
// to production.
func (s *Site) HasSlotBindings() bool {
	for _, b := range s.Bindings {
		if b.Slot != "" {
			return true
		}
	}
	return false
}

// SlotSite derives the configuration a slot's instances run with and its
// deployments are made with: production's, with the slot's release,
// variables and instance count. Bindings are left as they are (all of the
// site's): the process manager only asks whether the site serves HTTP,
// which must not differ between slots. It returns nil when the site has
// no such slot. The result shares no mutable state with s.
func SlotSite(s *Site, slot string) *Site {
	sl := s.FindSlot(slot)
	if sl == nil || s.Node == nil {
		return nil
	}
	out := *s
	n := *s.Node
	out.Node = &n
	out.Slot = slot
	out.ActiveRelease = sl.ActiveRelease
	out.Node.Env = SlotEnv(s.Node.Env, sl.Env)
	if sl.Instances > 0 {
		out.Node.Instances = sl.Instances
	}
	out.Slots, out.Tasks = nil, nil // tasks run in production only
	return &out
}

// SwapSite is the configuration a slot's instances are restarted with
// before a swap: exactly production's, running the slot's release. After
// the swap production is given that same configuration, so the warm
// instances carry on without a restart.
func SwapSite(s *Site, slot string) *Site {
	out := *s
	out.ActiveRelease = s.ReleaseIn(slot)
	return &out
}

// SlotEnv merges a slot's variables into production's: production's
// variables that are slot settings are left out, the slot's replace those
// of the same name and the rest are appended in order.
func SlotEnv(prod, slot []EnvVar) []EnvVar {
	out := make([]EnvVar, 0, len(prod)+len(slot))
	for _, e := range prod {
		if !e.SlotSetting {
			out = append(out, e)
		}
	}
	for _, e := range slot {
		e.SlotSetting = false
		if i := slices.IndexFunc(out, func(o EnvVar) bool { return o.Name == e.Name }); i >= 0 {
			out[i] = e
		} else {
			out = append(out, e)
		}
	}
	return out
}

// StatusRange is an inclusive range of HTTP statuses.
type StatusRange struct{ From, To int }

// ParseStatusRanges reads "200-399" or "200-299,401,404".
func ParseStatusRanges(s string) ([]StatusRange, error) {
	var out []StatusRange
	for _, part := range strings.Split(s, ",") {
		part = strings.TrimSpace(part)
		if part == "" {
			continue
		}
		lo, hi, isRange := strings.Cut(part, "-")
		from, err1 := strconv.Atoi(strings.TrimSpace(lo))
		to := from
		var err2 error
		if isRange {
			to, err2 = strconv.Atoi(strings.TrimSpace(hi))
		}
		if err1 != nil || err2 != nil || from < 100 || to > 599 || from > to {
			return nil, fmt.Errorf("%q is not an HTTP status or a range of them (100-599)", part)
		}
		out = append(out, StatusRange{from, to})
	}
	if len(out) == 0 {
		return nil, fmt.Errorf("list at least one status, e.g. 200-399")
	}
	return out, nil
}

// StatusAccepted reports whether code falls in one of the ranges.
func StatusAccepted(code int, ranges []StatusRange) bool {
	for _, r := range ranges {
		if code >= r.From && code <= r.To {
			return true
		}
	}
	return false
}

func (w *WarmupConfig) applyDefaults() {
	if len(w.Paths) == 0 {
		w.Paths = []string{"/"}
	}
	for i := range w.Paths {
		w.Paths[i] = strings.TrimSpace(w.Paths[i])
	}
	if strings.TrimSpace(w.Statuses) == "" {
		w.Statuses = DefaultWarmupStatuses
	}
	if w.TimeoutSec <= 0 {
		w.TimeoutSec = DefaultWarmupTimeoutSec
	}
}

// applySlotDefaults normalizes slot names and the slot of each binding.
func (s *Site) applySlotDefaults() {
	for i := range s.Slots {
		sl := &s.Slots[i]
		sl.Name = strings.ToLower(strings.TrimSpace(sl.Name))
		sl.Warmup.applyDefaults()
	}
	for i := range s.Bindings {
		s.Bindings[i].Slot = NormalizeSlot(s.Bindings[i].Slot)
	}
}

// validateSlots checks the slots and what they need of the site: a node
// or worker site with automatic ports (two slots cannot share one fixed
// port), bindings that name an existing slot, sane warm-up settings.
func (s *Site) validateSlots() error {
	for i, b := range s.Bindings {
		if b.Slot != "" && s.FindSlot(b.Slot) == nil {
			return verr(fmt.Sprintf("bindings[%d].slot", i), "there is no deployment slot %q", b.Slot)
		}
	}
	if len(s.Slots) == 0 {
		return nil
	}
	if !s.RunsNode() {
		return verr("slots", "deployment slots need a Node.js application or background worker site")
	}
	if len(s.Slots) > MaxSlots {
		return verr("slots", "at most %d deployment slots besides production", MaxSlots)
	}
	if s.Node.PortMode == "fixed" {
		return verr("node.portMode", "deployment slots need automatic ports: two slots cannot listen on one fixed port")
	}
	seen := map[string]bool{}
	for i, sl := range s.Slots {
		f := fmt.Sprintf("slots[%d]", i)
		if sl.Name == ProductionSlot || !slotNameRe.MatchString(sl.Name) {
			return verr(f+".name", "use 1-32 lower-case letters, digits or '-', other than \"production\"")
		}
		if seen[sl.Name] {
			return verr(f+".name", "another slot is called %q", sl.Name)
		}
		seen[sl.Name] = true
		if sl.Instances < 0 || sl.Instances > 64 {
			return verr(f+".instances", "must be between 0 (as many as production) and 64")
		}
		names := map[string]bool{}
		for j, e := range sl.Env {
			if !envNameRe.MatchString(e.Name) {
				return verr(fmt.Sprintf("%s.env[%d].name", f, j), "%q is not a valid variable name", e.Name)
			}
			if names[e.Name] {
				return verr(fmt.Sprintf("%s.env[%d].name", f, j), "%s is set twice", e.Name)
			}
			names[e.Name] = true
		}
		w := sl.Warmup
		if len(w.Paths) > 10 {
			return verr(f+".warmup.paths", "at most 10 warm-up paths")
		}
		for j, p := range w.Paths {
			if !strings.HasPrefix(p, "/") || strings.ContainsAny(p, " \t\r\n") {
				return verr(fmt.Sprintf("%s.warmup.paths[%d]", f, j), "must be a path starting with /")
			}
		}
		if _, err := ParseStatusRanges(w.Statuses); err != nil {
			return verr(f+".warmup.statuses", "%v", err)
		}
		if w.TimeoutSec < 5 || w.TimeoutSec > MaxWarmupTimeoutSec {
			return verr(f+".warmup.timeoutSec", "must be between 5 and %d seconds", MaxWarmupTimeoutSec)
		}
	}
	return nil
}

// ---- status, swaps

// SlotStatus is one slot (production included) as the API shows it.
type SlotStatus struct {
	Name     string     `json:"name"` // "production" or the slot's name
	Release  string     `json:"release,omitempty"`
	Status   SiteStatus `json:"status"`
	Bindings []Binding  `json:"bindings"`
	AutoSwap bool       `json:"autoSwap,omitempty"`
}

// SlotsView is GET /sites/{id}/slots: every slot, production first, and
// the swap in progress or the last one.
type SlotsView struct {
	Slots    []SlotStatus  `json:"slots"`
	Swap     *SwapProgress `json:"swap,omitempty"`
	LastSwap *SwapResult   `json:"lastSwap,omitempty"`
}

// Phases of a swap.
const (
	SwapPreparing = "preparing" // the slot's instances restart with production's settings
	SwapWarming   = "warming"   // warm-up requests
	SwapSwapping  = "swapping"  // production traffic moves
)

// SwapProgress is a swap in progress.
type SwapProgress struct {
	Slot      string    `json:"slot"`
	Phase     string    `json:"phase"`
	Message   string    `json:"message,omitempty"`
	User      string    `json:"user,omitempty"`
	Auto      bool      `json:"auto,omitempty"` // started by auto-swap after a deployment
	StartedAt time.Time `json:"startedAt"`
}

// SwapResult is how the last swap of a site ended. It is kept in memory:
// the event log and the audit log have the history.
type SwapResult struct {
	Slot       string    `json:"slot"`
	Succeeded  bool      `json:"succeeded"`
	Message    string    `json:"message"`
	User       string    `json:"user,omitempty"`
	Auto       bool      `json:"auto,omitempty"`
	StartedAt  time.Time `json:"startedAt"`
	FinishedAt time.Time `json:"finishedAt"`
	// Releases after the swap (before it, when it failed).
	ProductionRelease string `json:"productionRelease,omitempty"`
	SlotRelease       string `json:"slotRelease,omitempty"`
}

// SwapPreview is what a swap would do, for the confirmation dialog.
type SwapPreview struct {
	Slot              string       `json:"slot"`
	ProductionRelease string       `json:"productionRelease"` // now in production; goes to the slot
	SlotRelease       string       `json:"slotRelease"`       // now in the slot; goes to production
	Warmup            WarmupConfig `json:"warmup"`
	// Changes lists what happens, in order, in plain words.
	Changes []string `json:"changes"`
	// Warnings are reasons to think twice; Blockers make the swap refuse.
	Warnings []string `json:"warnings"`
	Blockers []string `json:"blockers"`
}

File: web/src/lib/slots.ts
// Deployment slots: pure helpers mirroring internal/model/slots.go.

import type { Binding, Deployment, DeploymentSlot, EnvVar, InstanceStatus, Site, SiteStatus, SwapPhase, SwapProgress, SwapResult, SlotsView } from '@/api/types';
import { runsNode } from './siteDefaults';

/** Names the site itself where a slot name is expected. */
export const PRODUCTION = 'production';
/** How many slots a site may have besides production. */
export const MAX_SLOTS = 4;
export const SLOT_NAME_RE = /^[a-z0-9]([a-z0-9-]{0,30}[a-z0-9])?$/;
export const DEFAULT_WARMUP_STATUSES = '200-399';
export const DEFAULT_WARMUP_TIMEOUT_SEC = 120;
export const MIN_WARMUP_TIMEOUT_SEC = 5;
export const MAX_WARMUP_TIMEOUT_SEC = 1800;
export const MAX_WARMUP_PATHS = 10;

/** A new slot with the server's defaults. */
export function newSlot(name = 'staging'): DeploymentSlot {
  return { name, autoSwap: false, warmup: { paths: ['/'], statuses: DEFAULT_WARMUP_STATUSES, timeoutSec: DEFAULT_WARMUP_TIMEOUT_SEC } };
}

/** The slots of a node or worker site (none for other types). */
export function siteSlots(site: Pick<Site, 'type' | 'slots'> | null | undefined): DeploymentSlot[] {
  return site && runsNode(site.type) ? (site.slots ?? []) : [];
}

export function hasSlots(site: Pick<Site, 'type' | 'slots'> | null | undefined): boolean {
  return siteSlots(site).length > 0;
}

/** Display name of a slot: "" and undefined are production. */
export function slotLabel(name: string | undefined | null): string {
  return name ? name : PRODUCTION;
}

/** The server's form of a slot name: "production" is the site itself (""). */
export function normalizeSlot(name: string | undefined | null): string {
  const n = (name ?? '').trim().toLowerCase();
  return n === PRODUCTION ? '' : n;
}

/** Why a slot name cannot be used, or null. `existing` are the other slots' names. */
export function validateSlotName(name: string, existing: string[]): string | null {
  const n = name.trim();
  if (!n) return 'Enter a name';
  if (n.toLowerCase() === PRODUCTION) return '"production" is the site itself: pick another name';
  if (!SLOT_NAME_RE.test(n)) return "1–32 lower-case letters, digits or '-', not starting or ending with '-'";
  if (existing.includes(n)) return `Another slot is called ${n}`;
  return null;
}

/** A free name to suggest for a new slot. */
export function suggestSlotName(existing: string[]): string {
  for (const n of ['staging', 'testing', 'preview', 'canary']) if (!existing.includes(n)) return n;
  for (let i = 2; ; i++) if (!existing.includes(`staging-${i}`)) return `staging-${i}`;
}

// ---------------------------------------------------------------- warm-up

export interface StatusRange {
  from: number;
  to: number;
}

export type StatusRangesResult = { ok: true; ranges: StatusRange[] } | { ok: false; error: string };

const INT_RE = /^[+-]?\d+$/;

/** Reads "200-399" or "200-299,401,404" like the server (statuses 100..599). */
export function parseStatusRanges(s: string): StatusRangesResult {
  const ranges: StatusRange[] = [];
  for (const raw of s.split(',')) {
    const part = raw.trim();
    if (!part) continue;
    const dash = part.indexOf('-');
    const lo = (dash >= 0 ? part.slice(0, dash) : part).trim();
    const hi = dash >= 0 ? part.slice(dash + 1).trim() : lo;
    const from = INT_RE.test(lo) ? parseInt(lo, 10) : NaN;
    const to = INT_RE.test(hi) ? parseInt(hi, 10) : NaN;
    if (!Number.isFinite(from) || !Number.isFinite(to) || from < 100 || to > 599 || from > to) {
      return { ok: false, error: `"${part}" is not an HTTP status or a range of them (100-599)` };
    }
    ranges.push({ from, to });
  }
  if (ranges.length === 0) return { ok: false, error: 'List at least one status, e.g. 200-399' };
  return { ok: true, ranges };
}

/** The error of an accepted-statuses list, or null. */
export function statusRangesError(s: string): string | null {
  const r = parseStatusRanges(s);
  return r.ok ? null : r.error;
}

export function statusAccepted(code: number, ranges: StatusRange[]): boolean {
  return ranges.some((r) => code >= r.from && code <= r.to);
}

/** Why a warm-up path is not valid, or null. */
export function warmupPathError(p: string): string | null {
  if (!p.startsWith('/')) return 'Must start with /';
  if (/\s/.test(p)) return 'No spaces';
  return null;
}

export function warmupTimeoutError(sec: number): string | null {
  return sec < MIN_WARMUP_TIMEOUT_SEC || sec > MAX_WARMUP_TIMEOUT_SEC
    ? `Between ${MIN_WARMUP_TIMEOUT_SEC} and ${MAX_WARMUP_TIMEOUT_SEC} seconds`
    : null;
}

// ---------------------------------------------------------------- variables

/**
 * The variables a slot's instances run with, as the server merges them:
 * production's variables that are not slot settings, then the slot's own,
 * replacing those of the same name or appended in order.
 */
export function effectiveSlotEnv(prod: EnvVar[] | null | undefined, slot: EnvVar[] | null | undefined): EnvVar[] {
  const out: EnvVar[] = (prod ?? []).filter((e) => !e.slotSetting);
  for (const e of slot ?? []) {
    const v: EnvVar = { ...e };
    delete v.slotSetting;
    const i = out.findIndex((o) => o.name === v.name);
    if (i >= 0) out[i] = v;
    else out.push(v);
  }
  return out;
}

export type EnvOrigin = 'production' | 'overridden' | 'slot';

export interface SlotEnvSummary {
  /** The effective variables with where each comes from. */
  vars: { name: string; secret: boolean; origin: EnvOrigin }[];
  /** Production's slot settings: not given to the slot. */
  sticky: string[];
}

/** Where each of a slot's effective variables comes from. */
export function slotEnvSummary(prod: EnvVar[] | null | undefined, slot: EnvVar[] | null | undefined): SlotEnvSummary {
  const own = new Set((slot ?? []).map((e) => e.name));
  const inherited = new Set((prod ?? []).filter((e) => !e.slotSetting).map((e) => e.name));
  return {
    vars: effectiveSlotEnv(prod, slot).map((e) => ({
      name: e.name,
      secret: !!e.secret,
      origin: own.has(e.name) ? (inherited.has(e.name) ? 'overridden' : 'slot') : 'production',
    })),
    sticky: (prod ?? []).filter((e) => e.slotSetting && e.name).map((e) => e.name),
  };
}

// ---------------------------------------------------------------- releases

/** Whether a site's release id names this deployment. */
export function releaseIs(release: string | undefined | null, dep: Pick<Deployment, 'id' | 'releaseDir'>): boolean {
  if (!release) return false;
  return release === dep.id || (!!dep.releaseDir && dep.releaseDir.endsWith(release));
}

/** The slots ("production" first) that run this deployment's release. */
export function releaseSlots(site: Pick<Site, 'type' | 'activeRelease' | 'slots'>, dep: Pick<Deployment, 'id' | 'releaseDir'>): string[] {
  const out: string[] = [];
  if (releaseIs(site.activeRelease, dep)) out.push(PRODUCTION);
  for (const sl of siteSlots(site)) if (releaseIs(sl.activeRelease, dep)) out.push(sl.name);
  return out;
}

/** Which slot runs a release ("production" first), or null. */
export function releaseSlot(site: Pick<Site, 'type' | 'activeRelease' | 'slots'>, releaseId: string): string | null {
  return releaseSlots(site, { id: releaseId })[0] ?? null;
}

/** The release a slot runs ("" or "production" = the site's own). */
export function slotRelease(site: Pick<Site, 'type' | 'activeRelease' | 'slots'>, slot: string): string | undefined {
  const n = normalizeSlot(slot);
  if (!n) return site.activeRelease;
  return siteSlots(site).find((s) => s.name === n)?.activeRelease;
}

// ---------------------------------------------------------------- selects

/** Production ("") and each slot, for a select. */
export function slotOptions(site: Pick<Site, 'type' | 'slots'>, productionLabel = 'Production'): { value: string; label: string }[] {
  return [{ value: '', label: productionLabel }, ...siteSlots(site).map((s) => ({ value: s.name, label: s.name }))];
}

/** Bindings that route to a slot ("" = production). */
export function slotBindings(bindings: Binding[] | null | undefined, slot: string): Binding[] {
  const n = normalizeSlot(slot);
  return (bindings ?? []).filter((b) => (b.slot ?? '') === n);
}

// ---------------------------------------------------------------- editing

/** The site's slots and bindings without a slot (and the bindings that routed to it). */
export function withoutSlot(site: Pick<Site, 'slots' | 'bindings'>, name: string): Pick<Site, 'slots' | 'bindings'> {
  return {
    slots: (site.slots ?? []).filter((s) => s.name !== name),
    bindings: (site.bindings ?? []).filter((b) => (b.slot ?? '') !== name),
  };
}

// ---------------------------------------------------------------- status and swaps

export const SWAP_PHASES: SwapPhase[] = ['preparing', 'warming', 'swapping'];

const phaseLabels: Record<SwapPhase, string> = {
  preparing: 'Preparing',
  warming: 'Warming up',
  swapping: 'Swapping',
};

const phaseDescriptions: Record<SwapPhase, string> = {
  preparing: "The slot's instances restart with production's settings",
  warming: 'Warm-up requests to every instance',
  swapping: 'Production traffic moves to the warm instances',
};

export function swapPhaseLabel(phase: string | undefined): string {
  return phaseLabels[phase as SwapPhase] ?? (phase || 'Starting');
}

export function swapPhaseDescription(phase: string | undefined): string {
  return phaseDescriptions[phase as SwapPhase] ?? '';
}

/** Position of a phase in SWAP_PHASES, -1 when unknown. */
export function swapStepIndex(phase: string | undefined): number {
  return SWAP_PHASES.indexOf(phase as SwapPhase);
}

/**
 * How a swap that was in progress ended: the view's lastSwap once the swap
 * is gone, or null while it runs (or when the result is of another swap).
 */
export function finishedSwap(prev: SwapProgress | null | undefined, view: SlotsView | undefined): SwapResult | null {
  if (!prev || !view || view.swap) return null;
  const r = view.lastSwap;
  if (!r || r.slot !== prev.slot) return null;
  const started = Date.parse(prev.startedAt);
  const finished = Date.parse(r.finishedAt);
  if (Number.isFinite(started) && Number.isFinite(finished) && finished < started) return null;
  return r;
}

const live = (i: InstanceStatus) => i.state !== 'exited' && i.state !== 'crashed';

/** Ready and running instances of a status. */
export function instanceCounts(status: Pick<SiteStatus, 'instances'> | null | undefined): { ready: number; total: number } {
  const list = (status?.instances ?? []).filter(live);
  return { ready: list.filter((i) => i.state === 'ready').length, total: list.length };
}

// ---------------------------------------------------------------- logs

/** Log filter value for every slot's lines. */
export const ALL_SLOTS = '*';

/**
 * The `slot` parameter of a log request for a filter value (ALL_SLOTS, ""
 * for production or a slot's name). Application lines of every slot need
 * none; an access log is always one slot's, production's by default.
 */
export function logSlotParam(value: string, access: boolean): string | undefined {
  if (value === ALL_SLOTS) return access ? PRODUCTION : undefined;
  return value || PRODUCTION;
}
